src/category_theory/limits/cofinal.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.limits
+import category_theory.punit
+import category_theory.comma
+import category_theory.is_connected
+
+/-!
+# Cofinal functors
+
+A functor `F : C ⥤ D` is cofinal if for every `d : D`,
+the comma category of morphisms `d ⟶ F.obj c` is connected.
+
+We prove that when `F : C ⥤ D` is cofinal,
+a functor `G : D ⥤ E` has a colimit if and only if `F ⋙ D` does
+(and in either case, the colimits are isomorphic).
+
+There is a converse which we don't prove here.
+
+## Naming
+There is some discrepancy in the literature about naming; some say 'final' instead of 'cofinal'.
+The explanation for this is that the 'co' prefix here is *not* the usual category-theoretic one
+indicating duality, but rather indicating the sense of "along with".
+
+While the trend seems to be towards using 'final', for now we go with the bulk of the literature
+and use 'cofinal'.
+
+## References
+* https://stacks.math.columbia.edu/tag/09WN
+* https://ncatlab.org/nlab/show/final+functor
+* Borceux, Handbook of Categorical Algebra I, Section 2.11.
+  (Note he reverses the roles of definition and main result relative to here!)
+-/
+
+noncomputable theory
+
+universes v u
+
+namespace category_theory
+open category_theory.limits
+
+variables {C : Type v} [small_category C]
+variables {D : Type v} [small_category D]
+
+/--
+A functor `F : C ⥤ D` is cofinal if for every `d : D`, the comma category of morphisms `d ⟶ F.obj c`
+is connected.
+
+See https://stacks.math.columbia.edu/tag/04E6
+-/
+def cofinal (F : C ⥤ D) : Prop :=
+∀ (d : D), is_connected (comma (functor.from_punit d) F)
+
+attribute [class] cofinal
+
+instance (F : C ⥤ D) [ℱ : cofinal F] (d : D) : is_connected (comma (functor.from_punit d) F) :=
+ℱ d
+
+namespace cofinal
+
+variables (F : C ⥤ D) [ℱ : cofinal F]
+include ℱ
```

Comment:
Could you comment on why it's necessary to name this typeclass argument and include it?

Reply:
Apparently it's not necessary at all!

Reply:
Even better!